src/targets/go.ts

```diff
@@ -0,0 +1,327 @@
+   */
+  readonly version?: string;
+}
+
+/**
+ * Information about a specific module.
+ */
+export interface GoModule {
+
+  /**
+   * Path to the mod file.
+   */
+  readonly modFile: string;
+
+  /**
+   * The version of the module.
+   */
+  readonly version: string;
+
+  /**
+   * The cannonical name of the module. (e.g 'github.com/aws/constructs-go/constructs/v3`)
+   */
+  readonly cannonicalName: string;
+
+  /**
+   * The path inside the repository the module is located in. (e.g 'constructs')
+   */
+  readonly repoPath: string;
+
+  /**
+   * The repository URL. (e.g 'github.com/aws/constructs')
+   */
+  readonly repoURL: string;
+
+}
+
+/**
+ * Release a set of Golang modules.
+ */
+export class GoReleaser {
+
+  private readonly version?: string;
+
+  private readonly dir: string;
+  private readonly dryRun: boolean;
+  private readonly gitBranch: string;
+  private readonly gitUsername: string;
+  private readonly gitUseremail: string;
+
+  constructor(props: GoReleaserProps) {
+
+    if (!shell.which('git')) {
+      throw new Error('git must be available to in order to be able to push Go code to GitHub');
+    }
+
+    this.version = props.version;
+    this.dir = props.dir ?? path.join(process.cwd(), 'dist', 'go');
+    this.gitBranch = props.branch ?? 'main';
+    this.dryRun = props.dryRun ?? false;
+    this.gitUsername = props.username ?? git.username();
+    this.gitUseremail = props.email ?? git.email();
+
+    if (!this.gitUseremail) {
+      throw new Error('Unable to detect username. either configure a git user.name or pass GIT_USER_NAME env variable');
+    }
+
+    if (!this.gitUsername) {
+      throw new Error('Unable to detect user email. either configure a git user.email or pass GIT_USER_EMAIL env variable');
+    }
+  }
+
+  /**
+   * Run the release process.
+   *
+   * @returns metadata about the release.
+   */
+  public release(): GoRelease {
+    const modules = this.collectModules(this.dir);
+    if (modules.length === 0) {
+      console.log('No modules detected. Skipping');
+      return {};
+    }
+
+    console.log('Detected modules:');
+    modules.forEach(m => console.log(` - ${m.modFile}`));
+
+    const repoURL = this.extractRepoURL(modules);
+    const repoDir = path.join(os.mkdtempSync(), 'repo');
+    git.clone(repoURL, repoDir);
+
+    const cwd = process.cwd();
+    try {
+      process.chdir(repoDir);
+      return this.doRelease(modules, repoDir);
+    } finally {
+      process.chdir(cwd);
+    }
+
+  }
+
+  private doRelease(modules: GoModule[], repoDir: string): GoRelease {
+
+    git.identify(this.gitUsername, this.gitUseremail);
+    git.checkout(this.gitBranch, { create: true });
+    this.syncRepo(repoDir);
+
+    git.add('.');
+    if (!git.diffIndex()) {
+      console.log('No changes. Skipping release');
+      return {};
+    }
+
+    const commitMessage = process.env.GIT_COMMIT_MESSAGE ?? this.buildReleaseMessage(modules);
```

Comment:
Slightly weird that this read from an environment variable while most other options are passed through the api

Reply:
Good catch. missed that in the refactor 👍